src/main/common/kalman.c

```diff
@@ -49,22 +49,31 @@ void kalman_init(void)
     init_kalman(&kalmanFilterStateRate[Z],  gyroConfig()->imuf_yaw_q);
 }
 
-void update_kalman_covariance(float gyroRateData, int axis)
+void update_kalman_covariance(kalman_t *kalmanState, float rate)
 {
-     kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex] = gyroRateData;
-     kalmanFilterStateRate[axis].axisSumMean +=  kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex];
-     kalmanFilterStateRate[axis].axisSumVar =  kalmanFilterStateRate[axis].axisSumVar + ( kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex] *  kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex]);
-     kalmanFilterStateRate[axis].windex++;
-    if ( kalmanFilterStateRate[axis].windex >= kalmanFilterStateRate[axis].w)
-    {
-         kalmanFilterStateRate[axis].windex = 0;
-    }
-     kalmanFilterStateRate[axis].axisSumMean -=  kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex];
-     kalmanFilterStateRate[axis].axisSumVar =  kalmanFilterStateRate[axis].axisSumVar - ( kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex] *  kalmanFilterStateRate[axis].axisWindow[ kalmanFilterStateRate[axis].windex]);
-     kalmanFilterStateRate[axis].axisMean =  kalmanFilterStateRate[axis].axisSumMean *  kalmanFilterStateRate[axis].inverseN;
-     kalmanFilterStateRate[axis].axisVar =  fabsf(kalmanFilterStateRate[axis].axisSumVar *  kalmanFilterStateRate[axis].inverseN - ( kalmanFilterStateRate[axis].axisMean *  kalmanFilterStateRate[axis].axisMean));
-
-    kalmanFilterStateRate[axis].r = sqrtf(kalmanFilterStateRate[axis].axisVar) * VARIANCE_SCALE;
+  kalmanState->axisWindow[kalmanState->windex] = rate;
+
+  kalmanState->axisSumMean += kalmanState->axisWindow[kalmanState->windex];
+  float varianceElement = kalmanState->axisWindow[kalmanState->windex] - kalmanState->axisMean;
+  varianceElement = varianceElement * varianceElement;
+  kalmanState->axisSumVar += varianceElement;
+  kalmanState->varianceWindow[kalmanState->windex] = varianceElement;
+
+  kalmanState->windex++;
+  if (kalmanState->windex >= kalmanState->w) {
+    kalmanState->windex = 0;
+  }
+
+  kalmanState->axisSumMean -= kalmanState->axisWindow[kalmanState->windex];
+  kalmanState->axisSumVar -= kalmanState->varianceWindow[kalmanState->windex];
+
+  //New mean
+  kalmanState->axisMean = kalmanState->axisSumMean * kalmanState->inverseN;
+  kalmanState->axisVar = kalmanState->axisSumVar * kalmanState->inverseN;
```

Comment:
@DzikuVx , we don't need absolute value here?  
i.e. older-code has `axisVar = fabsf(...);` & helio-imuf-code has `[x|y|z]Var = ABS(...);`.
i'm asking these things because i'm trying to back-port this math-fix to Helio-imu-firmware, thank you very much.

Reply:
@nerdCopter not really as `varianceElement` is squared, so it's always a positive number.